io/src/lib.rs

```diff
@@ -67,6 +76,29 @@ impl<'a, R: Read + ?Sized> Read for Take<'a, R> {
     }
 }
 
+// Impl copied from Rust stdlib.
+impl<'a, R: BufRead + ?Sized> BufRead for Take<'a, R> {
+    #[inline]
+    fn fill_buf(&mut self) -> Result<&[u8]> {
+        // Don't call into inner reader at all at EOF because it may still block
+        if self.remaining == 0 {
+            return Ok(&[]);
+        }
+
+        let buf = self.reader.fill_buf()?;
+        let cap = cmp::min(buf.len() as u64, self.remaining) as usize;
+        Ok(&buf[..cap])
+    }
+
+    #[inline]
+    fn consume(&mut self, amount: usize) {
+        // Don't let callers reset the limit by passing an overlarge value.
+        let amount = cmp::min(amount as u64, self.remaining) as usize;
```

Comment:
I think a (debug) panic would be better. Passing large value is considered invalid anyway.

Reply:
I used `assert!` since the stdlib version panics if amount is too big - added a panics section to the  rustdocs of the trait also.

Reply:
> added a panics section to the rustdocs of the trait also.

I'm not sure it's needed since it's already documented to be invalid in the trait itself but doesn't hurt.